### ISSUE TITLE
Is there any proposal to add anything analogous to OCaml's `function` keyword or Haskell's `LambdaCase` extension?

### ISSUE THREAD

**Issue by alissa-tung** [rescript-lang/syntax#430]
In [there](https://rescript-lang.org/docs/manual/latest/pattern-matching-destructuring)
```rescript
let optionBoolToBool = opt => {
  switch opt {
  | Some(trueOrFalse) => trueOrFalse
  | None => false
  }
}
```
seems redundant. Since there is no need for reference to `opt`.

**Comment by amiralies** [comment]
We used to have this in reason syntax with `fun` keyword.
The reason behind removing this was reckless currying. @IwanKaramazow can explain better.
I'm also missing this as i write more code in rescript but at the same time I don't like the `fun` syntax in reason.
I was thinking about about something like scala's case match expression.
```scala
 val f = {
   case Some(b) => b
   case None => false
}
```
so in rescript it would look something like this
```rescript
let f = {
| Some(b) => b
| None => false
}
```
or 
```rescript
let f = switch {
| Some(b) => b
| None => false
}
```
but they are ambigious and may be hard to parse.

it may be good to think about it, there's also an issue when converting from ml/reason which can be solved if we introduce this feature #147 .

**Comment by hoichi** [comment]
I was thinking about repurposing `switch` for this lately. Hopefully, it could also work inline. Because sometimes instead of this:

```res
switch
  value
  ->someOperation(a)
  ->someOtherOperation(b, c)
{
  | Some(x) => ...
  | None => ...
}
```

You really want this:

```res
  value
  ->someOperation(a)
  ->someOtherOperation(b, c)
  ->switch {
    | Some(x) => ...
	| None => ...
	}
```

**Comment by amiralies** [comment]
I like the `switch _`  syntax.
